File: HackerRank/Algorithms/Implementation/Sherlock and the Beast/sherlock_and_the_beast.py
def largest_decent(n):
    if n < 3:
        return -1

    # Try greedy; start with all 5's and see if it's valid.
    curr = int("".join(['5'] * n))

    three_size = 0

    while not is_decent(curr):
        # Threes have to be multiples of 5 so add 5 at a time.
        three_size += 5

        if three_size <= n:
            threes = "".join(['3'] * three_size)
            fives = "".join(['5'] * (n - three_size))

            curr = int(fives + threes)
        else:
            break

    # By the end of this, if it's not decent, then there is no valid number.
    if not is_decent(curr):
        return -1

    return curr

def is_decent(n):
    threes = 0
    fives = 0

    for digit in str(n):
        # Only 3 or 5's can be its digits.
        if digit != '3' and digit != '5':
            return False

        # Count 3's and 5's for divisibility.
        if digit == '3':
            threes += 1
        elif digit == '5':
            fives += 1

    # Once the threes and fives have been counted, make sure they're divisible.
    return threes % 5 == 0 and fives % 3 == 0

File: HackerRank/Algorithms/Implementation/Sherlock and the Beast/sherlock_and_the_beast.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;

bool isDecent(string& number) {
    int three = 0;
    int five = 0;
    
    for (auto it = number.begin(); it != number.end(); ++it) {
        // 1. Its digits can only be 3's and/or 5's.
        if (*it != '3' && *it != '5') {
            return false;
        }
        // Count 3's and 5's for bullets 2 and 3.
        if (*it == '3') {
            ++three;
        } else if (*it == '5') {
            ++five;
        }
    }
    
    /*
    2. The number of 3's it contains is divisible by 5.
    3. The number of 5's it contains is divisible by 3.
    */
    if (three % 5 == 0 && five % 3 == 0) {
        return true;
    }
    
    // If all else fails, it's false.
    return false;
}

string largestDecent(int digitCount) {
    string result = string(digitCount, '5');
    
    // Each iteration, check to see if the result is valid. If it is, done.
    for (int i = result.length() - 1; i >= 0; --i) {
        if (isDecent(result)) {
            return result;
        }
        
        // If the current number was not decent, then change last digit to 3.
        result[i] = '3';
    }
    
    // Since the case where the swapping at index 0 isn't evaluated, do it manually.
    if (isDecent(result)) {
        return result;
    }
    
    // If nothing was decent, then return -1.
    return "-1";
}

int main(){
    int t;
    cin >> t;
    for(int a0 = 0; a0 < t; a0++){
        int n;
        cin >> n;
        cout << largestDecent(n) << endl;
    }
    return 0;
}
